dart/common/Addon.h

```diff
@@ -0,0 +1,305 @@
+ *   AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *   LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *   ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *   POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef DART_COMMON_ADDON_H_
+#define DART_COMMON_ADDON_H_
+
+#include <string>
+
+#include "dart/common/Extensible.h"
+#include "dart/common/detail/NoOp.h"
+
+namespace dart {
+namespace common {
+
+class AddonManager;
+
+class Addon
+{
+public:
+
+  friend class AddonManager;
+
+  /// If your Addon has a State class, then that State class should inherit this
+  /// Addon::State class. This allows us to safely serialize, store, and clone
+  /// the states of arbitrary Addon extensions. If your Addon is stateless, then
+  /// you do not have to worry about extending this class, because
+  /// Addon::getState() will simply return a nullptr by default, which is taken
+  /// to indicate that it is stateless.
+  ///
+  /// The distinction between the State class and the Properties class is that
+  /// State will get stored in AddonManager::State whereas Properties will get
+  /// stored in AddonManager::Properties. Typically Properties are values that
+  /// only change rarely if ever, whereas State contains values that might
+  /// change as often as every time step.
+  class State : public Extensible<State> { };
+
+  /// Use the StateMixer class to easily create a State extension from an
+  /// existing class or struct.
+  template <class Mixin>
+  using StateMixer = ExtensibleMixer<State, Mixin>;
+
+  /// If your Addon has a Properties class, then it should inherit this
+  /// Addon::Properties class. This allows us to safely serialize, store, and
+  /// clone the properties of arbitrary Addon extensions. If your Addon has no
+  /// properties, then you do not have to worry about extending this class,
+  /// because Addon::getProperties() will simply return a nullptr by default,
+  /// which is taken to indicate that it has no properties.
+  ///
+  /// The distinction between the State class and the Properties class is that
+  /// State will get stored in AddonManager::State whereas Properties will get
+  /// stored in AddonManager::Properties. Typically Properties are values that
+  /// only change rarely if ever, whereas State contains values that might
+  /// change as often as every time step.
+  class Properties : public Extensible<Properties> { };
+
+  /// Use the PropertiesMixer class to easily create a Properties extension
+  /// from an existing class or struct.
+  template <class Mixin>
+  using PropertiesMixer = ExtensibleMixer<Properties, Mixin>;
+
+  /// Virtual destructor
+  virtual ~Addon() = default;
+
+  /// Clone this Addon into a new manager
+  virtual std::unique_ptr<Addon> cloneAddon(AddonManager* newManager) const = 0;
+
+  /// Set the State of this Addon. By default, this does nothing.
+  virtual void setAddonState(const std::unique_ptr<State>& otherState);
```

Comment:
This should take a const reference, not a `std::unique_ptr`. The current signature prevents me from passing a stack-allocated `State` or a state managed by `shared_ptr` as an argument. In general, [it's *never* necessary to pass a `std::unique_ptr` by const reference](http://www.codingstandard.com/rule/8-2-4-do-not-pass-stdunique_ptr-by-const-reference/) because you can always pass the contents by reference, instead.

This comment also applies to `setAddonProperties`.